Massdrop/fluxible-router
# Gjp/better scroll position tracking

Note: hold on merging for a bit, this needs significant testing to be sure it works in all browsers/devices

At time of writing, this branch is live at https://psyduck.massdrop.com

#### Asana Ticket
https://app.asana.com/0/107988962060869/138775396468648
https://app.asana.com/0/107988962060869/144345846574286

#### Description of changes
Rewrote scroll position tracking to save state on navigation event, not on scroll. This resolves History API limitations introduced in recent Safari (desktop and mobile) versions without needing UA sniffing checks (and may also be good for general page performance, since we aren't querying scrollX/Y so often and forcing reflows). This fix also improves scroll position tracking during really fast backwards/forwards navigation, where the previous implementation would occasionally fail to properly preserve scroll state if pages are navigated between quickly enough (due to the interplay between the 150ms timer and an "isNavigationComplete" check).

This fix might introduce a regression in one instance: if a browser is forcefully terminated (so unload handlers don't run), changes to the scroll state won't be preserved.

This included adding incrementing ids to history session states, for distinguishing back events from forward events (the native "popstate" event doesn't reveal that information).

#### Steps to test
1. Go to any page (p1)
2. Scroll the page to some specific point (y1) and remember it (I recommend lining up edges of the browser with distinctive page elements, like images/dropcards/etc.)
3. Click on a link to navigate to another page (p2)
4. Repeat step 2 (scroll to some point y2)
5. Click on a link to navigate to another page (p3)
6. Repeat step 3 (scroll to some point y3)
7. Click the back button, you should return to page p2, scrolled to point y2
8. Click the back button, you should return to page p1, scrolled to point y1
9. Click the forward button, you should return to page p2, scrolled to point y2
10. Click the forward button, you should return to page p3, scrolled to point y3

Note when testing with Massdrop: you may see "jumping" where, after some navigation, the page renders scrolled at the very top and after that returns to the saved scroll position. This seems to be due to our page content being temporarily emptied, and is not a regression relative to master on prod

#### Known Issue

(Specific to Massdrop): going to a page that itself auto-scrolls to some location (direct link to a comment for example), that auto-scroll overrides restoring the saved scroll state. This is not a regression relative to previous functionality.


---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/massdrop/fluxible-router/10)
